casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala

```diff
@@ -126,47 +126,66 @@ object EstimatorHelper {
     * @param sc
     * @return
     */
-  def computeMergeChanges[F[_]: Sync, C, P, A, K](
+  def computeMergeChanges[F[_]: Concurrent, C, P, A, K](
       historyRepo: HistoryRepository[F, C, P, A, K],
       baseState: Blake2b256Hash,
       mainDeploys: List[ProcessedDeploy],
       mergingDeploys: List[ProcessedDeploy]
-  )(implicit sc: Serialize[C]): F[MergeChanges] = {
+  ): F[MergeChanges] = {
+
     // errored deploy is always non-conflict
-    val mainNonErrDeploys    = mainDeploys.filter(!_.isFailed)
-    val mergingNonErrDeploys = mergingDeploys.filter(!_.isFailed)
+    val successDeploysMain  = mainDeploys.filterNot(_.isFailed)
+    val successDeploysMerge = mergingDeploys.filterNot(_.isFailed)
+
+    // free IOEvents are produces and consumes that are not part of any COMM event, or they are persistent
+    def getFreeIOEvents(d: ProcessedDeploy) = {
+      val (prod, cons, comms) = d.deployLog
+        .map(EventConverter.toRspaceEvent)
+        .foldLeft((Seq.empty[Produce], Seq.empty[Consume], Seq.empty[COMM])) { (op, event) =>
+          event match {
+            case p: Produce => (op._1 :+ p, op._2, op._3)
+            case c: Consume => (op._1, op._2 :+ c, op._3)
+            case c: COMM    => (op._1, op._2, op._3 :+ c)
+          }
+        }
+      val freeProd = prod.filter(p => p.persistent || !comms.exists(_.produces.contains(p)))
```

Comment:
In this case, peek would be ignored.

If a peek matched, the comm event would be like Seq(COMM(Consume(X), Produce(X)), Consume(X), Produce(X), Produce(X)), which there should a free Prod left.
But the filter ignore this situation.

Reply:
I think I do not understand - why peek produces that event log? If peek = produce following consume, why event log is not `Seq(COMM(Consume(X), Produce(X)), Consume(X), Produce(X)`?

Reply:
Try Rholang like 
```
@0!(1)|
for (_ <<- @0) { 0 }
```
And see what is the eventLogs.

 My question is -> Is there supposed to be a free produce in this situation?

Reply:
I think yes, there should be free produce, because it has not been consumed. It just been read (peeked), which should effectively be noop for tuplespace.

Reply:
> free produce here will be part of comm event. Right?
Yes

Reply:
I think the problem here is how peek is implemented, it should not mess up with produces and consumes.

Reply:
But sadly, this is currently how we are doing now.